gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/RefreshTokenAuth2Provider.java

```diff
@@ -0,0 +1,231 @@
+              + "&client_id="
+              + clientId
+              + "&client_secret="
+              + clientSecret.value();
+      conn.getOutputStream().write(payload.getBytes("UTF-8"));
+
+      int httpResponseCode = conn.getResponseCode();
+      logger.atFine().log("Response " + httpResponseCode);
+
+      String responseContentType = conn.getHeaderField("Content-Type");
+
+      if (httpResponseCode == HttpURLConnection.HTTP_OK
+          && responseContentType.startsWith("application/json")) {
+        logger.atFine().log("Received a 200 with presumably a new access token");
+        InputStream httpResponseStream = conn.getInputStream();
+        token = parseTokenFromStream(httpResponseStream);
+      } else {
+        logger.atFine().log("Refresh token grant flow failed");
+        InputStream stream = conn.getErrorStream();
+        if (stream == null) {
+          // no error stream, try the original input stream
+          stream = conn.getInputStream();
+        }
+        String responseBody = consumeInputStream(stream);
+        logger.atSevere().log(
+            "Received following response '"
+                + responseBody
+                + "' with code '"
+                + httpResponseCode
+                + "'");
+      }
+    } finally {
+      if (conn != null) {
+        conn.disconnect();
+      }
+    }
+    return token;
+  }
+
+  private String consumeInputStream(InputStream stream) throws IOException {
+    BufferedReader br = new BufferedReader(new InputStreamReader(stream, "utf-8"));
+    StringBuilder response = new StringBuilder();
+    String responseLine;
+    while ((responseLine = br.readLine()) != null) {
+      response.append(responseLine.trim());
+    }
+    return response.toString();
+  }
+
+  private AccessToken parseTokenFromStream(InputStream httpResponseStream) throws IOException {
+    AccessToken token;
+    try {
+      int expirationPeriodInSecs = 0;
+
+      JsonFactory jf = new JsonFactory();
+      JsonParser jp = jf.createJsonParser(httpResponseStream);
+      String fieldName, fieldValue;
+      jp.nextToken();
+      String accessTokenFromResponse = "";
+      while (jp.hasCurrentToken()) {
+        if (jp.getCurrentToken() == JsonToken.FIELD_NAME) {
+          fieldName = jp.getCurrentName();
+          jp.nextToken(); // field value
+          fieldValue = jp.getText();
+
+          if (fieldName.equals("access_token")) {
+            accessTokenFromResponse = fieldValue;
+          }
+
+          if (fieldName.equals("expires_in")) {
+            expirationPeriodInSecs = Integer.parseInt(fieldValue);
+          }
+        }
+        jp.nextToken();
+      }
+
+      token =
+          new AccessToken(
+              accessTokenFromResponse,
+              System.currentTimeMillis() + (expirationPeriodInSecs * 1000L));
+      jp.close();
+    } finally {
+      httpResponseStream.close();
+    }
+    return token;
+  }
+
+  public Proxy getProxy(String proxyAddress) {
```

Comment:
Can we make this method `private` ?

Reply:
Wont be application once I use the google client instead.